build-runtime-host.sh

```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env bash
+
```

Comment:
I realize this method of "declaring a host module" comes from the approach in KEwasm. But thinking about it, I believe we can improve on that method a bit. As it stands, we either need to write a Wasm host module the way this script does (which is frail), or write a K definition of the module (which is larger than it needs to be).

Instead of declaring a new instruction for all host functions and wrapping the host functions in a module, we could extend the Wasm semantics. By specifying which modules should be considered host modules (in this case `"env"`, in the Ewasm case `"ethereum"`) we could add rules that say something to the effect of:

```
syntax Set ::= "#HostModules" [function]
syntax Instr ::= #hostCall(WasmSting, WasmString, TypeDecl)

rule (import MODNAME FUNCNAME (func OID:OptionalId T:TypeDecl)) => (func OID T (#hostCall(MODNAME, FUNCNAME, T))
  requires MODNAME in #HostModules
```

So the rule for the `phost . FNAME` would instead be for `#hostCall(MODNAME, FNAME)`, it could be made simpler, and there would be a rule `rule #HostModules => SetItem(#unparseWasmString("\"env\""))`

Not saying this is something we should do now, necessarily, but it's something we should open an issue about,because it would make defining new embeddings easier. The concept of a host call would be native to the Wasm semantics, and each embedding would really only need to define the `#HostModules` set and each `#hostCall` it cares about.

Reply:
One thing I don't like about this approach is that it puts the host functions into the same module that is being currently defined, instead of into a separate Wasm module like the `import` statement expects. Not entirely sure this will cause any problems, but it definitely means different code paths will be executed.

But we could definitely have a mode where we "auto-construct" any implicit imported modules, so that any `(import "MODULE_NAME" ...)` statements automatically just add the corresponding statement to the module registry in the appropriate place (instead of to the current module).

The current approach is pretty low overhead though, and I don't think it's that frail. The `KWASM-POLKADOT-HOST` K module is pretty minimal, and the code this script is working over is auto-generated, so it should be uniform enough to be post-processed. But we won't know how frail it actually is until we try it out a bunch. I just tried to codify the process I was using locally to do the updates as a starting point here.

Reply:
That's a good point. So the best approach would be to allocate missing imports as host imports into the store directly. The good thing is that they don't get reallocated if they are imported in more than one module. And since imports are handled first in a module, it's basically equivalent to pre-allocating them, which is the correct approach.

The frailty comes from it relying on imports and type declarations being declared on one line each. If `wasm2wat` changes its pretty-printing that breaks. It's not something we need to rush to fix. But since there is a way to deal with this directly in the semantics which also reflectes the official-spec way to do things, I think it's worth fixing at some point.

I think another selling point is that it makes it easier to implement new embeddings. I'd prefer if it wasn't necessary to either set up a script like the one above or to define a file like `ewasm.md` in the `ewasm-semantics` repo. With an auto-construct mode, if no host function is called, the regular kwasm runner will work with no changes. If a host function is called, the runner will halt when it reaches it, which is the same as not returning control. If the host function should do something, the behavior can be added as a semantic rule. So an "embedder" file would just be a list of relevant host function rules.